addons/delivery/models/sale_order.py

```diff
@@ -9,20 +9,22 @@
 class SaleOrder(models.Model):
     _inherit = 'sale.order'
 
-    pickup_location_data = fields.Json()
     carrier_id = fields.Many2one('delivery.carrier', string="Delivery Method", check_company=True, help="Fill this field if you plan to invoice the shipping based on picking.")
+    is_pickup_delivery = fields.Boolean(related='carrier_id.is_pickup')
     delivery_message = fields.Char(readonly=True, copy=False)
     delivery_set = fields.Boolean(compute='_compute_delivery_state')
     recompute_delivery_price = fields.Boolean('Delivery cost should be recomputed')
     is_all_service = fields.Boolean("Service Product", compute="_compute_is_service_products")
     shipping_weight = fields.Float("Shipping Weight", compute="_compute_shipping_weight", store=True, readonly=False)
 
-    def _compute_partner_shipping_id(self):
-        """ Override to reset the delivery address when a pickup location was selected. """
-        super()._compute_partner_shipping_id()
-        for order in self:
-            if order.partner_shipping_id.is_pickup_location:
-                order.partner_shipping_id = order.partner_id
+    def write(self, vals):
+        if 'carrier_id' in vals:
+            new_carrier = self.env['delivery.carrier'].browse(vals['carrier_id'])
+            if any(order.partner_shipping_id.pickup_delivery_carrier_id and order.partner_shipping_id.pickup_delivery_carrier_id.id != new_carrier.id for order in self):
+                orders_with_wrong_partner = self.filtered(lambda so: so.partner_shipping_id.pickup_delivery_carrier_id and so.partner_shipping_id.pickup_delivery_carrier_id.id != new_carrier.id)
+                for order in orders_with_wrong_partner:
+                    order.partner_shipping_id = order.partner_id
+        return super().write(vals)
```

Comment:
I think it is better to always reset in case a pickup location is outdated. We want to force the choice of it through the selector.
```suggestion
    def write(self, vals):
        """Reset the shipping partner when it is a pickup location and the delivery method changes."""
        if 'carrier_id' in vals:
            for order in self:
                if order.partner_shipping_id.pickup_delivery_carrier_id:
                    order.partner_shipping_id = order.partner_id
        return super().write(vals)
```